drivers/java/src/main/java/com/rethinkdb/net/SocketWrapper.java

```diff
@@ -2,47 +2,86 @@
 
 import com.rethinkdb.gen.exc.ReqlDriverError;
 
+import javax.net.SocketFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
 import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
 import java.net.InetSocketAddress;
+import java.net.Socket;
 import java.net.SocketException;
 import java.net.SocketTimeoutException;
 import java.nio.ByteBuffer;
-import java.nio.channels.SocketChannel;
-import java.net.StandardSocketOptions;
+import java.nio.ByteOrder;
 import java.nio.charset.StandardCharsets;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
 import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.ReentrantLock;
 
 public class SocketWrapper {
-    private SocketChannel socketChannel;
-    private Optional<Long> timeout = Optional.empty();
+    final ReentrantLock mu = new ReentrantLock();
+
+    private Socket socket = null;
+
+    private SocketFactory socketFactory = SocketFactory.getDefault();
+    private SSLSocket sslSocket = null;
```

Comment:
Actually, we should leave it as is since `sslSocket` wraps `socket`. Besides `connect(...)` we only use `socket` on method `close()`. Hence, no need for null checking.